src/main/java/com/softserve/bookstore/repositories/UserRepository.java

```diff
@@ -2,123 +2,148 @@
 import com.softserve.bookstore.models.Role;
 import com.softserve.bookstore.models.User;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Repository;
 import org.tinylog.Logger;
 
-import java.io.IOException;
 import java.sql.Connection;
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
-import java.util.Collections;
-import java.util.List;
-import java.util.Optional;
-
-//read about life of beans in spring boot
-//automatic connection creation in spring
-//read about repositories in spring
-//services, repositories, components, how does it works inside spring..
-//database initialisation for spring
-//!!!!!
-//CTRL ALT L -format
+import java.util.*;
+
 @Repository
-public class UserRepository implements RepositoryInterface<User> {
+public class UserRepository {
 
     public static final String INSERT_USERS = "INSERT INTO users(email, password) VALUES (?,?)";
     public static final String SELECT_USERS = "SELECT * FROM users";
     public static final String SELECT_LAST_USERS = "SELECT TOP (?) * FROM users ORDER BY id_user DESC ";
+    public static final String INSERT_USERS_ROLES = "INSERT INTO users_roles(id_user, id_role) VALUES(?,?)";
     public static final String INSERT_ORDERS = "INSERT INTO orders(id_user, date, status) VALUES(?,?,?)";
 
     @Autowired
-    ManageUserData manageUserData;
-
-    @Autowired
-    ConnectionManager connectionManager;
-
-    @Autowired
-    OrderRepository orderRepository;
-
-    @Autowired
-    RoleRepository roleRepository;
-
-
-    public List<User> findAllFromFile(String fileName) throws IOException {
-        return manageUserData.readDataFromFile(fileName);
-    }
+    private ConnectionManager connectionManager;
 
     public List<User> findAll() throws SQLException {
         Connection connection = connectionManager.getConnection();
         PreparedStatement userStatement = connection.prepareStatement(SELECT_USERS);
         ResultSet resultSet = userStatement.executeQuery();
         Logger.info("Users were successfully retrived from the database.");
-        return UserUtility.getUsersFromResultSet(resultSet);
+        return getUsersFromResultSet(resultSet);
     }
 
+    public List<User> findLastUsersAdded(int numberOfRecords) throws SQLException {
+        Connection connection = connectionManager.getConnection();
+        PreparedStatement userStatement = connection.prepareStatement(SELECT_LAST_USERS);
+        userStatement.setInt(1, numberOfRecords);
+        ResultSet resultSet = userStatement.executeQuery();
+        List<User> users = getUsersFromResultSet(resultSet);
+        Collections.reverse(users);
+        return users;
+    }
+
+    private  void addUserToBatch(PreparedStatement userStatement, User user) {
+        try {
+            userStatement.setString(1, user.getEmail());
+            userStatement.setString(2, user.getPassword());
+            userStatement.addBatch();
```

Comment:
For exception handling just have single point where exception will be handled, 
in most cases it is done at the level of controller layer (web service layer), where exception will be handled
exception message will be included in response with proper response, like 404, 503, etc.
and also logging should be done at that point of exception handling

Reply:
FYI: here is the article which help you understand some ways it can be done in spring application
take a look especially at the @ControllerAdvice annotation.
https://www.baeldung.com/exception-handling-for-rest-with-spring
but for now let's just keep it simple and have exception handling at web service layer